Sacado: "declaration shadows a typedef" warnings from clang

@trilinos/sacado 

## Expectations
Builds with clang that use the `-Wshadow` flag should not emit warnings.

## Current Behavior
Two such warnings are emitted by clang when building `Sacado_trad.hpp`.
```
.../include/Sacado_trad.hpp:1469:32: warning: declaration shadows a typedef in 'ADcontext<Double>' [-Wshadow]
        typedef ADvari<Double> ADVari;
                               ^
.../include/Sacado_trad.hpp:234:32: note: previous declaration is here
        typedef ADvari<Double> ADVari;
                               ^
.../include/Sacado_trad.hpp:1508:32: warning: declaration shadows a typedef in 'ADcontext<Double>' [-Wshadow]
        typedef ADvari<Double> ADVari;
                               ^
.../include/Sacado_trad.hpp:234:32: note: previous declaration is here
        typedef ADvari<Double> ADVari;
```

Another warning is emitted by clang when building `Kokkos_ViewFactory.hpp`:
```
.../include/Kokkos_ViewFactory.hpp:82:11: warning: declaration shadows a typedef in 'ViewFactory<ViewPack...>' [-Wshadow]
    using value_type = typename ResultView::non_const_value_type;
          ^
.../include/Kokkos_ViewFactory.hpp:74:61: note: previous declaration is here
  typedef typename Impl::ViewFactoryType<ViewPack...>::type value_type;
```

## Possible Solution
For `Sacado_trad.hpp`, since the shadowing declaration is identical to the shadowed one, likely the correct fix is just to delete the shadowing one (lines 1508 and 1469 of `Sacado_trad.hpp`).  For `Kokkos_ViewFactory.hpp`, it is not as obvious to me what the correct fix is.

## Environment
I'm building on a Mac using Apple clang 10.0.0.

--- a/packages/phalanx/example/FiniteElementAssembly/evaluators/Constant.hpp
+++ b/packages/phalanx/example/FiniteElementAssembly/evaluators/Constant.hpp
@@ -42,7 +42,7 @@
 // @HEADER
 
 
-#ifndef PHX_EXAMPLE_EVALAUTOR_CONSTANT_HPP
+#ifndef PHX_EXAMPLE_EVALUATOR_CONSTANT_HPP
 #define PHX_EXAMPLE_EVALUATOR_CONSTANT_HPP
 
 #include "Phalanx_Evaluator_WithBaseImpl.hpp"

--- a/packages/phalanx/example/FiniteElementAssembly/evaluators/ProjectGradientToQP.hpp
+++ b/packages/phalanx/example/FiniteElementAssembly/evaluators/ProjectGradientToQP.hpp
@@ -42,7 +42,7 @@
 // @HEADER
 
 #ifndef PHX_PROJECT_GRADIENT_TO_QP_HPP
-#define PHX_PROEJCT_GRADIENT_TO_QP_HPP
+#define PHX_PROJECT_GRADIENT_TO_QP_HPP
 
 #include "Phalanx_Evaluator_WithBaseImpl.hpp"
 #include "Phalanx_Evaluator_Derived.hpp"

--- a/packages/phalanx/example/FiniteElementAssembly/evaluators/ProjectValueToQP.hpp
+++ b/packages/phalanx/example/FiniteElementAssembly/evaluators/ProjectValueToQP.hpp
@@ -42,7 +42,7 @@
 // @HEADER
 
 #ifndef PHX_PROJECT_VALUE_TO_QP_HPP
-#define PHX_PROEJCT_VALUE_TO_QP_HPP
+#define PHX_PROJECT_VALUE_TO_QP_HPP
 
 #include "Phalanx_Evaluator_WithBaseImpl.hpp"
 #include "Phalanx_Evaluator_Derived.hpp"

--- a/packages/phalanx/example/FiniteElementAssembly_DeviceDag/evaluators/Constant.hpp
+++ b/packages/phalanx/example/FiniteElementAssembly_DeviceDag/evaluators/Constant.hpp
@@ -42,7 +42,7 @@
 // @HEADER
 
 
-#ifndef PHX_EXAMPLE_EVALAUTOR_CONSTANT_HPP
+#ifndef PHX_EXAMPLE_EVALUATOR_CONSTANT_HPP
 #define PHX_EXAMPLE_EVALUATOR_CONSTANT_HPP
 
 #include "Phalanx_Evaluator_WithBaseImpl.hpp"

--- a/packages/phalanx/example/FiniteElementAssembly_DeviceDag/evaluators/ProjectGradientToQP.hpp
+++ b/packages/phalanx/example/FiniteElementAssembly_DeviceDag/evaluators/ProjectGradientToQP.hpp
@@ -42,7 +42,7 @@
 // @HEADER
 
 #ifndef PHX_PROJECT_GRADIENT_TO_QP_HPP
-#define PHX_PROEJCT_GRADIENT_TO_QP_HPP
+#define PHX_PROJECT_GRADIENT_TO_QP_HPP
 
 #include "Phalanx_Evaluator_WithBaseImpl.hpp"
 #include "Phalanx_Evaluator_Derived.hpp"

--- a/packages/phalanx/example/FiniteElementAssembly_DeviceDag/evaluators/ProjectValueToQP.hpp
+++ b/packages/phalanx/example/FiniteElementAssembly_DeviceDag/evaluators/ProjectValueToQP.hpp
@@ -42,7 +42,7 @@
 // @HEADER
 
 #ifndef PHX_PROJECT_VALUE_TO_QP_HPP
-#define PHX_PROEJCT_VALUE_TO_QP_HPP
+#define PHX_PROJECT_VALUE_TO_QP_HPP
 
 #include "Phalanx_Evaluator_WithBaseImpl.hpp"
 #include "Phalanx_Evaluator_Derived.hpp"

--- a/packages/phalanx/example/FiniteElementAssembly_DeviceDag/evaluators/ScatterResidual.hpp
+++ b/packages/phalanx/example/FiniteElementAssembly_DeviceDag/evaluators/ScatterResidual.hpp
@@ -104,7 +104,7 @@ public:
                   const int& in_num_equations,
                   const Kokkos::View<const int**,PHX::Device>& in_gids);
   PHX::DeviceEvaluator<Traits>* createDeviceEvaluator() const override;
-  void evaluateFields(typename Traits::EvalData d);
+  void evaluateFields(typename Traits::EvalData d) override;
 };
 
 // **************************************************************
@@ -146,7 +146,7 @@ public:
                   const int& in_num_equations,
                   const Kokkos::View<const int**,PHX::Device>& in_gids);
   PHX::DeviceEvaluator<Traits>* createDeviceEvaluator() const override;
-  void evaluateFields(typename Traits::EvalData d);
+  void evaluateFields(typename Traits::EvalData d) override;
 };
 
 /*

--- a/packages/phalanx/example/FiniteElementAssembly_MixedFieldTypes/evaluators/Constant.hpp
+++ b/packages/phalanx/example/FiniteElementAssembly_MixedFieldTypes/evaluators/Constant.hpp
@@ -42,7 +42,7 @@
 // @HEADER
 
 
-#ifndef PHX_EXAMPLE_EVALAUTOR_CONSTANT_HPP
+#ifndef PHX_EXAMPLE_EVALUATOR_CONSTANT_HPP
 #define PHX_EXAMPLE_EVALUATOR_CONSTANT_HPP
 
 #include "Phalanx_Evaluator_WithBaseImpl.hpp"

--- a/packages/phalanx/example/FiniteElementAssembly_MixedFieldTypes/evaluators/ProjectGradientToQP.hpp
+++ b/packages/phalanx/example/FiniteElementAssembly_MixedFieldTypes/evaluators/ProjectGradientToQP.hpp
@@ -42,7 +42,7 @@
 // @HEADER
 
 #ifndef PHX_PROJECT_GRADIENT_TO_QP_HPP
-#define PHX_PROEJCT_GRADIENT_TO_QP_HPP
+#define PHX_PROJECT_GRADIENT_TO_QP_HPP
 
 #include "Phalanx_Evaluator_WithBaseImpl.hpp"
 #include "Phalanx_Evaluator_Derived.hpp"

--- a/packages/phalanx/example/FiniteElementAssembly_MixedFieldTypes/evaluators/ProjectValueToQP.hpp
+++ b/packages/phalanx/example/FiniteElementAssembly_MixedFieldTypes/evaluators/ProjectValueToQP.hpp
@@ -42,7 +42,7 @@
 // @HEADER
 
 #ifndef PHX_PROJECT_VALUE_TO_QP_HPP
-#define PHX_PROEJCT_VALUE_TO_QP_HPP
+#define PHX_PROJECT_VALUE_TO_QP_HPP
 
 #include "Phalanx_Evaluator_WithBaseImpl.hpp"
 #include "Phalanx_Evaluator_Derived.hpp"

--- a/packages/phalanx/src/Phalanx_FieldManager.hpp
+++ b/packages/phalanx/src/Phalanx_FieldManager.hpp
@@ -333,7 +333,7 @@ namespace PHX {
 
   private:
 
-    typedef PHX::EvaluationContainer_TemplateManager<Traits> SCTM;
+    using SCTM = PHX::EvaluationContainer_TemplateManager<Traits>;
 
     std::size_t m_num_evaluation_types;
 

--- a/packages/phalanx/src/Phalanx_FieldManager_Def.hpp
+++ b/packages/phalanx/src/Phalanx_FieldManager_Def.hpp
@@ -220,7 +220,6 @@ void PHX::FieldManager<Traits>::
 aliasFieldForAllEvaluationTypes(const PHX::FieldTag& aliasedField,
                                 const PHX::FieldTag& targetField)
 {
-  typedef PHX::EvaluationContainer_TemplateManager<Traits> SCTM;
   typename SCTM::iterator it = m_eval_containers.begin();
   for (; it != m_eval_containers.end(); ++it)
     it->aliasField(aliasedField,targetField);
@@ -242,8 +241,6 @@ inline
 void PHX::FieldManager<Traits>::
 requireFieldForAllEvaluationTypes(const PHX::FieldTag& t)
 {
-  typedef PHX::EvaluationContainer_TemplateManager<Traits> SCTM;
-
   typename SCTM::iterator it = m_eval_containers.begin();
   for (; it != m_eval_containers.end(); ++it) {
     it->requireField(t);
@@ -266,8 +263,6 @@ inline
 void PHX::FieldManager<Traits>::
 registerEvaluatorForAllEvaluationTypes(const Teuchos::RCP<PHX::Evaluator<Traits> >& e)
 {
-  typedef PHX::EvaluationContainer_TemplateManager<Traits> SCTM;
-
   typename SCTM::iterator it = m_eval_containers.begin();
   for (; it != m_eval_containers.end(); ++it) {
     it->registerEvaluator(e);
@@ -311,7 +306,6 @@ inline
 void PHX::FieldManager<Traits>::
 postRegistrationSetup(typename Traits::SetupData d, const bool& buildDeviceDAG)
 {
-  typedef PHX::EvaluationContainer_TemplateManager<Traits> SCTM;
   typename SCTM::iterator it = m_eval_containers.begin();
   for (std::size_t i = 0; it != m_eval_containers.end(); ++it, ++i)
     it->postRegistrationSetup(d, *this, buildDeviceDAG);
@@ -438,7 +432,6 @@ writeGraphvizFile(const std::string base_filename,
 		  bool writeDependentFields,
 		  bool debugRegisteredEvaluators) const
 {
-  typedef PHX::EvaluationContainer_TemplateManager<Traits> SCTM;
   typename SCTM::const_iterator it = m_eval_containers.begin();
   for (; it != m_eval_containers.end(); ++it) {
     std::string name = base_filename + "_" + it->evaluationType() +
@@ -453,7 +446,6 @@ template<typename Traits>
 inline
 void PHX::FieldManager<Traits>::print(std::ostream& os) const
 {
-  typedef PHX::EvaluationContainer_TemplateManager<Traits> SCTM;
   typename SCTM::const_iterator it = m_eval_containers.begin();
   for (; it != m_eval_containers.end(); ++it)
     os << (*it);

--- a/packages/phalanx/test/EvaluatorMacros/EvaluatorWithMacros_Def.hpp
+++ b/packages/phalanx/test/EvaluatorMacros/EvaluatorWithMacros_Def.hpp
@@ -46,9 +46,6 @@
 #include "Phalanx_DataLayout_MDALayout.hpp"
 #include "Phalanx_FieldTag_Tag.hpp"
 
-struct CELL;
-struct BASIS;
-
 namespace PHX {
 
   // Implementation #1

--- a/packages/phalanx/test/UnmanagedFields/MDField/MDField_TestEvaluators_Def.hpp
+++ b/packages/phalanx/test/UnmanagedFields/MDField/MDField_TestEvaluators_Def.hpp
@@ -46,8 +46,8 @@
 #include "Phalanx_DataLayout_MDALayout.hpp"
 #include "Phalanx_FieldTag_Tag.hpp"
 
-struct CELL;
-struct BASIS;
+class CELL;
+class BASIS;
 
 namespace PHX {
 

--- a/packages/phalanx/test/Utilities/Evaluator_MockDAG_Def.hpp
+++ b/packages/phalanx/test/Utilities/Evaluator_MockDAG_Def.hpp
@@ -46,8 +46,8 @@
 #include "Phalanx_DataLayout_MDALayout.hpp"
 #include "Phalanx_FieldTag_Tag.hpp"
 
-struct CELL;
-struct BASIS;
+class CELL;
+class BASIS;
 
 namespace PHX {
 

--- a/packages/sacado/src/Kokkos_ViewFactory.hpp
+++ b/packages/sacado/src/Kokkos_ViewFactory.hpp
@@ -79,8 +79,8 @@ struct ViewFactory {
               const CtorProp& prop,
               const Dims ... dims) {
 
-    using value_type = typename ResultView::non_const_value_type;
-    constexpr bool is_scalar = Sacado::IsScalarType<value_type>::value;
+    using nc_value_type = typename ResultView::non_const_value_type;
+    constexpr bool is_scalar = Sacado::IsScalarType<nc_value_type>::value;
     constexpr bool is_dyn_rank = is_dyn_rank_view<ResultView>::value;
 
     // rank == number of arguments

--- a/packages/sacado/src/Sacado_trad.hpp
+++ b/packages/sacado/src/Sacado_trad.hpp
@@ -1466,7 +1466,6 @@ template<typename Double> void ADcontext<Double>::do_init()
 
 template<typename Double> void ADcontext<Double>::free_all()
 {
-        typedef ADvari<Double> ADVari;
         typedef ConstADvari<Double> ConstADVari;
         ADMemblock *mb, *mb1;
 
@@ -1505,7 +1504,6 @@ template<typename Double> void ADcontext<Double>::free_all()
 
 template<typename Double> void ADcontext<Double>::re_init()
 {
-        typedef ADvari<Double> ADVari;
         typedef ConstADvari<Double> ConstADVari;
 
         if (ConstADVari::cadc.Busy || ADVari::adc.Busy || ADVari::adc.Free

